File: version_1/appcenter_uploader.rb
require 'net/http'
require 'uri'
require 'json'

module AppCenter
  class BaseAPI
    BASE_URL = 'https://api.appcenter.ms/v0.1/apps/'

    def initialize(owner_name, app_name, api_token)
      @owner_name = owner_name
      @app_name = app_name
      @api_token = api_token
    end

    def call_api(url, request)
      http = Net::HTTP.new(url.host, url.port)
      http.use_ssl = true

      response = http.request(request)

      if response.is_a?(Net::HTTPSuccess)
        JSON.parse(response.body)
      else
        puts "Error: #{response.code} - #{response.message}"
      end
    end
  end

  class ReleaseUploader < BaseAPI
    def upload_release
      url = URI("#{BASE_URL}#{@owner_name}/#{@app_name}/uploads/releases")
      request = Net::HTTP::Post.new(url)
      request['accept'] = 'application/json'
      request['X-API-Token'] = @api_token
      request['Content-Type'] = 'application/json'
      request['Content-Length'] = '0'

      call_api(url, request)
    end
  end

  class MetadataUploader < BaseAPI
    FILE_SIZE_COMMAND = "wc -c $RELEASE_FILE_LOCATION | awk '{print $1}'"

    def initialize(response)
      super(response.owner_name, response.app_name, response.api_token)
      @response = response
    end

    def upload_metadata(file_name, content_type)
      file_size_bytes = `#{FILE_SIZE_COMMAND}`.strip
      metadata_url = URI("#{BASE_URL}#{@owner_name}/#{@app_name}/uploads/releases/#{@response.id}")
      request = Net::HTTP::Post.new(metadata_url)
      request['Content-Type'] = 'application/json'
      request['Accept'] = 'application/json'
      request.body = {
        file_name: file_name,
        file_size: file_size_bytes,
        token: @response.url_encoded_token,
        content_type: content_type
      }.to_json

      call_api(metadata_url, request)
    end
  end

  class ChunkUploader < BaseAPI
    CHUNK_SIZE = 4194304 # 4 MB

    def upload_chunks(file_path)
      chunk_number = 0
      Dir.foreach(file_path) do |file|
        next if file == '.' || file == '..'

        chunk_number += 1
        content_length = File.size(file)
        upload_chunk_url = URI("https://file.appcenter.ms/upload/upload_chunk/#{@response.package_asset_id}?token=#{URI.encode_www_form_component(@response.url_encoded_token)}&block_number=#{chunk_number}")
        request = Net::HTTP::Post.new(upload_chunk_url)
        request.body = File.read(file)
        request['Content-Length'] = content_length.to_s
        request['Content-Type'] = 'application/octet-stream'

        call_api(upload_chunk_url, request)
      end
    end
  end

  class FinishUploader < BaseAPI
    def finish_upload
      finish_url = URI("https://file.appcenter.ms/upload/finished/#{@response.package_asset_id}?token=#{URI.encode_www_form_component(@response.url_encoded_token)}")
      request = Net::HTTP::Post.new(finish_url)

      call_api(finish_url, request)
    end
  end

  class StatusChecker < BaseAPI
    def check_status
      status_url = URI("#{BASE_URL}#{@owner_name}/#{@app_name}/uploads/releases/#{@response.id}")
      request = Net::HTTP::Get.new(status_url)

      call_api(status_url, request)
    end
  end

  class ReleaseDistributor < BaseAPI
    def distribute_release(distribution_group)
      distribute_url = URI("#{BASE_URL}#{@owner_name}/#{@app_name}/releases/#{@response.release_id}")
      request = Net::HTTP::Patch.new(distribute_url)
      request.body = { destinations: [{ name: distribution_group }] }.to_json
      request['Content-Type'] = 'application/json'
      request['Accept'] = 'application/json'
      request['X-API-Token'] = @api_token

      call_api(distribute_url, request)
    end
  end
end

File: version_1/appcenter_uploader.rs
use reqwest::blocking::{Client, RequestBuilder};
use std::fs;
use std::path::Path;

pub struct AppCenterAPI {
    base_url: String,
    owner_name: String,
    app_name: String,
    api_token: String,
}

impl AppCenterAPI {
    pub fn new(owner_name: &str, app_name: &str, api_token: &str) -> Self {
        Self {
            base_url: String::from("https://api.appcenter.ms/v0.1/apps/"),
            owner_name: owner_name.to_string(),
            app_name: app_name.to_string(),
            api_token: api_token.to_string(),
        }
    }

    pub fn upload_release(&self) -> Result<(), reqwest::Error> {
        let url = format!(
            "{}{}/{}/uploads/releases",
            self.base_url, self.owner_name, self.app_name
        );
        let client = Client::new();
        let request = client.post(&url).header("X-API-Token", &self.api_token);
        let response = request.send()?;
        // Handle response accordingly
        Ok(())
    }
}

pub struct MetadataUploader {
    base_url: String,
    response: Response,
}

impl MetadataUploader {
    pub fn new(response: Response) -> Self {
        Self {
            base_url: String::from("https://file.appcenter.ms/upload/set_metadata/"),
            response,
        }
    }

    pub fn upload_metadata(
        &self,
        file_name: &str,
        content_type: &str,
    ) -> Result<(), reqwest::Error> {
        let file_size_bytes = fs::metadata(file_name)?.len();
        let metadata_url = format!("{}/{}", self.base_url, self.response.package_asset_id);
        let client = Client::new();
        let request = client.post(&metadata_url).json(&json!({
            "file_name": file_name,
            "file_size": file_size_bytes,
            "token": self.response.url_encoded_token,
            "content_type": content_type
        }));
        let response = request.send()?;
        // Handle response accordingly
        Ok(())
    }
}

// Define other structs and their methods similarly

fn main() {
    // Example usage
    let api = AppCenterAPI::new("Example-Org", "Example-App", "Example-Token");
    api.upload_release().unwrap();

    let response = Response {
        id: "{ID}".to_string(),
        package_asset_id: "{PACKAGE_ASSET_ID}".to_string(),
        upload_domain: "https://file.appcenter.ms".to_string(),
        token: "{TOKEN}".to_string(),
        url_encoded_token: "{URL_ENCODED_TOKEN}".to_string(),
    };
    let uploader = MetadataUploader::new(response);
    uploader.upload_metadata("ExampleApp.apk", "application/vnd.android.package-archive").unwrap();
}

